Microsoft.Toolkit/Collections/ObservableGroupedCollectionExtensions.cs

```diff
@@ -268,16 +379,53 @@ public static class ObservableGroupedCollectionExtensions
         /// <param name="key">The key of the group where the item at <paramref name="index"/> should be removed.</param>
         /// <param name="index">The index of the item to remove in the group.</param>
         /// <param name="removeGroupIfEmpty">If true (default value), the group will be removed once it becomes empty.</param>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="index"/> is less than zero or <paramref name="index"/> is greater than the group elements' count.</exception>
         public static void RemoveItemAt<TKey, TValue>(
             this ObservableGroupedCollection<TKey, TValue> source,
             TKey key,
             int index,
             bool removeGroupIfEmpty = true)
+        {
+            if (source.TryGetList(out var list))
+            {
+                var groupIndex = 0;
+                foreach (var group in list!)
+                {
+                    if (EqualityComparer<TKey>.Default.Equals(group.Key, key))
+                    {
+                        group.RemoveAt(index);
+
+                        if (removeGroupIfEmpty && group.Count == 0)
+                        {
+                            source.RemoveAt(groupIndex);
+                        }
+
+                        return;
+                    }
+
+                    groupIndex++;
+                }
+            }
+            else
+            {
+                RemoveItemAtWithLinq(source, key, index, removeGroupIfEmpty);
+            }
+        }
+
+        /// <summary>
+        /// Slow path for <see cref="RemoveItemAt{TKey,TValue}"/>.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void RemoveItemAtWithLinq<TKey, TValue>(
+            ObservableGroupedCollection<TKey, TValue> source,
+            TKey key,
+            int index,
+            bool removeGroupIfEmpty)
         {
             var groupIndex = 0;
             foreach (var group in source)
             {
-                if (GroupKeyPredicate(group, key))
+                if (EqualityComparer<TKey>.Default.Equals(group.Key, key))
```

Comment:
Could we achieve the same performances by keeping `GroupKeyPredicate(group, key)` here and adding the `[MethodImpl(MethodImplOptions.AggressiveInlining)]` to the `GroupKeyPredicate()` method ?

The code feel easier to read with a named method

Reply:
Generally the JIT should be able to produce the same code at the end, yeah. Though also from a style perspective, personally I feel like just leaving `EqualityComparer<T>` call directly makes the code clearer in its behavior, instead of hiding the logic away. Like, while working on this PR I found myself looking up that method multiple times as I wasn't sure exactly what its semantics was or how it was working exactly. Since it's just a wrapper on this call and it's only shorter by a few characters, I think I'd prefer to just leave the `EqualityComparer<T>` invocation explicit. Just my two cents though, so let me know what you think! 😊

Reply:
I usually prefer having a named method to describe the intent (even if it here, it seems that it missed the point ☺). I find it make the code easier to maintain because we have a single update point if one day we change the equality logic (even if it here it is unlikely to happen).

Reply:
I see, yeah that makes sense. I guess I just tend to do the opposite in general out of habit when the wrapped code is so short just to give the JIT a litte nudge, but either approach is fine, really 😊

Not sure what you mean by "even if it here, it seems that it missed the point", would you like me to revert that change then and reintroduce that wrapping method, or do we want to leave this as is for now? Let me know and I'll make the changes if needed!

Reply:
> even if it here, it seems that it missed the point;

I was meaning that the naming wasn't that good since you said you had to look several times at the implementation to understand what it was doing 🙂.

We can leave this code as-is for now. No need to push another update only for this.

Reply:
Alright, perfect then! 😊
Thank you for the review!